File: src/lib/image-processor.ts
import sharp from 'sharp';
import path from 'path';
import fs from 'fs';

const UPLOAD_DIR = path.join(process.cwd(), 'public', 'uploads');

// 绿色背景（#00FF00）的 HSV 范围，用于色键抠图
const GREEN_MIN = { h: 60, s: 40, v: 20 };
const GREEN_MAX = { h: 180, s: 100, v: 100 };

/**
 * RGB 转 HSV
 */
function rgbToHsv(r: number, g: number, b: number): { h: number; s: number; v: number } {
  const rN = r / 255;
  const gN = g / 255;
  const bN = b / 255;
  const max = Math.max(rN, gN, bN);
  const min = Math.min(rN, gN, bN);
  const diff = max - min;

  let h = 0;
  if (diff !== 0) {
    if (max === rN) h = ((gN - bN) / diff + (gN < bN ? 6 : 0)) * 60;
    else if (max === gN) h = ((bN - rN) / diff + 2) * 60;
    else h = ((rN - gN) / diff + 4) * 60;
  }

  const s = max === 0 ? 0 : (diff / max) * 100;
  const v = max * 100;

  return { h, s, v };
}

/**
 * 移除图片中的绿色背景，转为透明 PNG
 * @param inputPath 输入图片路径（相对于 public/ 或绝对路径）
 * @param tolerance 容差值 0-100，越大移除越多绿色，默认 30
 * @returns 输出图片路径（相对于 public/）
 */
export async function removeGreenBackground(
  inputPath: string,
  tolerance: number = 30
): Promise<string> {
  // 解析输入路径
  let absInput: string;
  if (inputPath.startsWith('/uploads/')) {
    absInput = path.join(process.cwd(), 'public', inputPath);
  } else if (path.isAbsolute(inputPath)) {
    absInput = inputPath;
  } else {
    absInput = path.join(process.cwd(), 'public', inputPath);
  }

  if (!fs.existsSync(absInput)) {
    throw new Error(`Input file not found: ${absInput}`);
  }

  // 生成输出路径
  const ext = path.extname(absInput);
  const baseName = path.basename(absInput, ext);
  const outputName = `${baseName}_transparent.png`;
  const absOutput = path.join(path.dirname(absInput), outputName);

  // 读取图片像素数据
  const { data, info } = await sharp(absInput)
    .raw()
    .ensureAlpha()
    .toBuffer({ resolveWithObject: true });

  const pixelCount = info.width * info.height;
  const pixels = new Uint8Array(data);

  // 遍历每个像素，将绿色区域设为透明
  for (let i = 0; i < pixelCount; i++) {
    const offset = i * 4;
    const r = pixels[offset];
    const g = pixels[offset + 1];
    const b = pixels[offset + 2];

    const hsv = rgbToHsv(r, g, b);

    // 判断是否为绿色背景（带容差）
    const isGreen =
      hsv.h >= GREEN_MIN.h - tolerance &&
      hsv.h <= GREEN_MAX.h + tolerance &&
      hsv.s >= GREEN_MIN.s &&
      hsv.v >= GREEN_MIN.v;

    if (isGreen) {
      // 绿色像素 → 完全透明
      pixels[offset + 3] = 0;
    }
  }

  // 边缘柔化：对半透明边缘做 feather 处理
  const feathered = await featherEdges(pixels, info.width, info.height);

  // 输出为 PNG
  await sharp(feathered, {
    raw: {
      width: info.width,
      height: info.height,
      channels: 4,
    },
  })
    .png()
    .toFile(absOutput);

  // 返回相对路径
  const relativePath = `/uploads/${outputName}`;
  return relativePath;
}

/**
 * 边缘柔化处理：对透明与不透明边界做 1px 的 alpha 渐变
 */
async function featherEdges(
  pixels: Uint8Array,
  width: number,
  height: number
): Promise<Uint8Array> {
  const result = new Uint8Array(pixels);

  for (let y = 1; y < height - 1; y++) {
    for (let x = 1; x < width - 1; x++) {
      const idx = (y * width + x) * 4;
      const alpha = pixels[idx + 3];

      // 如果当前像素不透明，检查周围是否有透明像素
      if (alpha === 255) {
        let hasTransparentNeighbor = false;
        for (let dy = -1; dy <= 1; dy++) {
          for (let dx = -1; dx <= 1; dx++) {
            const nIdx = ((y + dy) * width + (x + dx)) * 4;
            if (pixels[nIdx + 3] === 0) {
              hasTransparentNeighbor = true;
              break;
            }
          }
          if (hasTransparentNeighbor) break;
        }

        // 边缘像素降低 alpha，做 1px 渐变
        if (hasTransparentNeighbor) {
          result[idx + 3] = 128;
        }
      }
    }
  }

  return result;
}

/**
 * 将 Sprite 图切割为单独的帧
 * @param inputPath sprite 图路径
 * @param rows 行数
 * @param cols 列数
 * @param outputPrefix 输出文件名前缀
 * @returns 切割后的帧路径列表
 */
export async function extractFrames(
  inputPath: string,
  rows: number,
  cols: number,
  outputPrefix?: string
): Promise<string[]> {
  let absInput: string;
  if (inputPath.startsWith('/uploads/')) {
    absInput = path.join(process.cwd(), 'public', inputPath);
  } else if (path.isAbsolute(inputPath)) {
    absInput = inputPath;
  } else {
    absInput = path.join(process.cwd(), 'public', inputPath);
  }

  if (!fs.existsSync(absInput)) {
    throw new Error(`Input file not found: ${absInput}`);
  }

  const metadata = await sharp(absInput).metadata();
  const imgWidth = metadata.width || 0;
  const imgHeight = metadata.height || 0;

  if (imgWidth === 0 || imgHeight === 0) {
    throw new Error('Invalid image dimensions');
  }

  const frameWidth = Math.floor(imgWidth / cols);
  const frameHeight = Math.floor(imgHeight / rows);

  const prefix = outputPrefix || `frame_${Date.now()}`;
  const frames: string[] = [];

  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      const left = col * frameWidth;
      const top = row * frameHeight;
      const outputName = `${prefix}_${row}_${col}.png`;
      const absOutput = path.join(UPLOAD_DIR, outputName);

      await sharp(absInput)
        .extract({ left, top, width: frameWidth, height: frameHeight })
        .png()
        .toFile(absOutput);

      frames.push(`/uploads/${outputName}`);
    }
  }

  return frames;
}

/**
 * 确保上传目录存在
 */
export function ensureUploadDir(): void {
  if (!fs.existsSync(UPLOAD_DIR)) {
    fs.mkdirSync(UPLOAD_DIR, { recursive: true });
  }
}

File: backend/services/image_processor.py
"""
Image processing service using Pillow.
Handles green screen removal and sprite frame extraction.
"""
import os
import time
import random
import string
from PIL import Image

UPLOAD_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "uploads")


def ensure_upload_dir():
    os.makedirs(UPLOAD_DIR, exist_ok=True)


def _rgb_to_hsv(r: int, g: int, b: int) -> tuple[float, float, float]:
    """Convert RGB (0-255) to HSV (h: 0-360, s: 0-100, v: 0-100)."""
    rn, gn, bn = r / 255.0, g / 255.0, b / 255.0
    mx = max(rn, gn, bn)
    mn = min(rn, gn, bn)
    diff = mx - mn

    h = 0.0
    if diff != 0:
        if mx == rn:
            h = ((gn - bn) / diff + (6 if gn < bn else 0)) * 60
        elif mx == gn:
            h = ((bn - rn) / diff + 2) * 60
        else:
            h = ((rn - gn) / diff + 4) * 60

    s = 0.0 if mx == 0 else (diff / mx) * 100
    v = mx * 100
    return h, s, v


def _resolve_path(path: str) -> str:
    """Resolve a URL path like /uploads/xxx.png to absolute filesystem path."""
    if path.startswith("/uploads/"):
        return os.path.join(UPLOAD_DIR, path[len("/uploads/"):])
    if os.path.isabs(path):
        return path
    return os.path.join(UPLOAD_DIR, path)


def remove_green_background(input_path: str, tolerance: int = 30) -> str:
    """
    Remove green background from image, output transparent PNG.
    Returns the /uploads/ relative URL.
    """
    abs_input = _resolve_path(input_path)
    if not os.path.exists(abs_input):
        raise FileNotFoundError(f"Input file not found: {abs_input}")

    ensure_upload_dir()

    img = Image.open(abs_input).convert("RGBA")
    pixels = img.load()
    width, height = img.size

    # Green HSV range
    green_min_h, green_max_h = 60, 180
    green_min_s, green_min_v = 40, 20

    for y in range(height):
        for x in range(width):
            r, g, b, a = pixels[x, y]
            h, s, v = _rgb_to_hsv(r, g, b)

            is_green = (
                green_min_h - tolerance <= h <= green_max_h + tolerance
                and s >= green_min_s
                and v >= green_min_v
            )
            if is_green:
                pixels[x, y] = (r, g, b, 0)

    # Feather edges: soften 1px boundary between opaque and transparent
    for y in range(1, height - 1):
        for x in range(1, width - 1):
            r, g, b, a = pixels[x, y]
            if a == 255:
                has_transparent_neighbor = False
                for dy in (-1, 0, 1):
                    for dx in (-1, 0, 1):
                        if pixels[x + dx, y + dy][3] == 0:
                            has_transparent_neighbor = True
                            break
                    if has_transparent_neighbor:
                        break
                if has_transparent_neighbor:
                    pixels[x, y] = (r, g, b, 128)

    # Save output
    base = os.path.splitext(os.path.basename(abs_input))[0]
    output_name = f"{base}_transparent.png"
    abs_output = os.path.join(UPLOAD_DIR, output_name)
    img.save(abs_output, "PNG")

    return f"/uploads/{output_name}"


def extract_frames(input_path: str, rows: int, cols: int, output_prefix: str | None = None) -> list[str]:
    """
    Cut a sprite sheet into individual frames.
    Returns list of /uploads/ relative URLs.
    """
    abs_input = _resolve_path(input_path)
    if not os.path.exists(abs_input):
        raise FileNotFoundError(f"Input file not found: {abs_input}")

    ensure_upload_dir()

    img = Image.open(abs_input)
    img_width, img_height = img.size

    if img_width == 0 or img_height == 0:
        raise ValueError("Invalid image dimensions")

    frame_width = img_width // cols
    frame_height = img_height // rows

    prefix = output_prefix or f"frame_{int(time.time())}"
    frames = []

    for row in range(rows):
        for col in range(cols):
            left = col * frame_width
            top = row * frame_height
            right = left + frame_width
            bottom = top + frame_height

            frame = img.crop((left, top, right, bottom))
            output_name = f"{prefix}_{row}_{col}.png"
            abs_output = os.path.join(UPLOAD_DIR, output_name)
            frame.save(abs_output, "PNG")
            frames.append(f"/uploads/{output_name}")

    return frames


def save_image_from_bytes(data: bytes, prefix: str = "gen") -> str:
    """Save raw image bytes to uploads dir, return /uploads/ relative URL."""
    ensure_upload_dir()
    ext = "png"
    unique = "".join(random.choices(string.ascii_lowercase + string.digits, k=6))
    filename = f"{prefix}_{int(time.time())}_{unique}.{ext}"
    filepath = os.path.join(UPLOAD_DIR, filename)
    with open(filepath, "wb") as f:
        f.write(data)
    return f"/uploads/{filename}"
